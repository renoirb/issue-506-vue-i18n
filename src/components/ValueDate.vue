<template>
  <time
    :datetime="datetime"
    :data-variant="variant"
    :data-value="value"
    :data-locale="locale"
    :data-alternate="alternate"
    :data-time-zone="timeZone"
  >
    <!-- or v-text="$d(value, variant, locale)" -->
    {{ $d(value, variant, locale) }}
  </time>
</template>

<script>
// @vue/component
export default {
  name: 'ValueDate',
  props: {
    value: {
      type: Number,
      required: true,
    },
    variant: {
      type: String,
      default: 'full',
    },
    timeZone: {
      type: String,
      default: 'America/Montreal',
    },
  },
  computed: {
    datetime() {
      const value = this.value
      return new Date(value).toISOString()
    },
    locale() {
      return this.$i18n.locale
    },
    alternate() {
      const value = this.value
      const variant = this.variant
      const locale = this.locale
      const formats = this.$i18n.getDateTimeFormat(locale)
      // console.log(`getDateTimeFormat(${locale})`, formats)
      const timeZone = this.timeZone
      const options = {
        timeZone,
        ...formats[variant],
      }
      return new Date(value).toLocaleString(locale, options)
    },
  },
}
</script>
