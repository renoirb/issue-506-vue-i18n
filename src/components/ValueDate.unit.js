import ValueDate from '@/components/ValueDate'

describe('ValueDate', () => {
  it('Exports a valid component', () => {
    expect(ValueDate).toBeAComponent()
  })

  it('Accept DateTimeFormat as argument', async () => {
    let variant = 'time'
    let locale = 'de-DE'

    const { localVue, i18n } = createComponentMocks({
      i18n: {
        locale,
      },
    })
    const value = 1550632495000
    const valueTime = '22:14:55'
    const datetime = '2019-02-20T03:14:55.000Z'

    const wrapper = mount(ValueDate, {
      propsData: { value, variant },
      localVue,
      i18n,
    })

    expect(wrapper.html()).toMatchSnapshot()
    expect(i18n.getDateTimeFormat(locale)).toMatchSnapshot()
    expect(wrapper.vm.value).toBe(value)
    expect(wrapper.text()).toBe(valueTime)
    expect(wrapper.attributes('datetime')).toBe(datetime)

    wrapper.setProps({ value, variant: 'long', timeZone: 'America/Fortaleza' })
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setProps({ value, variant: 'full', timeZone: 'Europe/Berlin' })
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()

    locale = 'fr-CA'
    i18n.locale = locale
    variant = 'fooBarBazz'
    const timeZone = 'America/Montreal'
    const anotherFormat = {
      fooBarBazz: {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
        hour: '2-digit',
        minute: '2-digit',
        timeZone,
      },
    }
    i18n.mergeDateTimeFormat(locale, anotherFormat)

    wrapper.setProps({ value, variant: 'long', timeZone })
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })
})
