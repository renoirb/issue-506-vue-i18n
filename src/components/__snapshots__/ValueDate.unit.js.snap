// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValueDate Accept DateTimeFormat as argument 1`] = `
<time datetime="2019-02-20T03:14:55.000Z" data-variant="time" data-value="1550632495000" data-locale="de-DE" data-alternate="22:14:55" data-time-zone="America/Montreal">
  22:14:55
</time>
`;

exports[`ValueDate Accept DateTimeFormat as argument 2`] = `
Object {
  "full": Object {
    "day": "numeric",
    "hour": "numeric",
    "hour12": false,
    "minute": "numeric",
    "month": "long",
    "second": "numeric",
    "timeZoneName": "long",
    "weekday": "long",
    "year": "numeric",
  },
  "long": Object {
    "day": "numeric",
    "month": "long",
    "weekday": "long",
    "year": "numeric",
  },
  "short": Object {
    "day": "2-digit",
    "hour": "2-digit",
    "minute": "2-digit",
    "month": "2-digit",
    "year": "numeric",
  },
  "shorter": Object {
    "day": "numeric",
    "month": "long",
    "year": "numeric",
  },
  "time": Object {
    "hour": "numeric",
    "hour12": false,
    "minute": "numeric",
    "second": "numeric",
  },
}
`;

exports[`ValueDate Accept DateTimeFormat as argument 3`] = `
<time datetime="2019-02-20T03:14:55.000Z" data-variant="long" data-value="1550632495000" data-locale="de-DE" data-alternate="Wednesday, February 20, 2019" data-time-zone="America/Fortaleza">
  Tuesday, February 19, 2019
</time>
`;

exports[`ValueDate Accept DateTimeFormat as argument 4`] = `
<time datetime="2019-02-20T03:14:55.000Z" data-variant="full" data-value="1550632495000" data-locale="de-DE" data-alternate="Wednesday, February 20, 2019, 04:14:55 Central European Standard Time" data-time-zone="Europe/Berlin">
  Tuesday, February 19, 2019, 22:14:55 GMT-05:00
</time>
`;

exports[`ValueDate Accept DateTimeFormat as argument 5`] = `
<time datetime="2019-02-20T03:14:55.000Z" data-variant="long" data-value="1550632495000" data-locale="fr-CA" data-alternate="Tuesday, February 19, 2019" data-time-zone="America/Montreal">
  Tuesday, February 19, 2019
</time>
`;
